sdk/iOS/src/MSManagedObjectObserver.m

```diff
@@ -7,35 +7,122 @@
 //
 
 #import "MSManagedObjectObserver.h"
+#import "MSCoreDataStore.h"
+#import "MSClient.h"
 #import <CoreData/CoreData.h>
 
 @interface MSManagedObjectObserver()
-@property (nonatomic, weak) MSClient *client;
+@property (nonatomic, strong) MSClient *client;
+@property (nonatomic, weak) NSManagedObjectContext *context;
 @end
 
 @implementation MSManagedObjectObserver
 
-- (id) initWithClient:(MSClient *)client
-{
-    return [self initWithClient:client contextsToObserver:nil];
-}
-
-- (id) initWithClient:(MSClient *)client contextsToObserver:(NSArray *)contexts
+- (instancetype) initWithClient:(MSClient *)client
 {
     self = [super init];
     if (self) {
-        _client = client;
+		// Copy so we can change the handling of how the sync table operations are handled
+        _client = [client copy];
+		
+		if ([_client.syncContext.dataSource isKindOfClass:[MSCoreDataStore class]])
+		{
+			MSCoreDataStore *dataStore = _client.syncContext.dataSource;
+			_context = dataStore.context;
+			
+			[[NSNotificationCenter defaultCenter] addObserver:self
+													 selector:@selector(handleDidSaveNotification:)
+														 name:NSManagedObjectContextDidSaveNotification
+													   object:_context];
+		}
+		
+		// Modify the handling of sync table operations for this instance of the data source
+		_client.syncContext.dataSource.handlesSyncTableOperations = FALSE;
     }
     return self;
 }
 
-- (void) registerAsObserver
+- (void)dealloc
+{
+	if (self.context != nil)
+	{
+		[[NSNotificationCenter defaultCenter] removeObserver:self
+														name:NSManagedObjectContextDidSaveNotification
+													  object:self.context];
+	}
+}
+
+/// Checks the entity has the ms_version attribute to signal it should
+/// be synchronised to the mobile service
+- (BOOL)entityHasRemoteStoreAttributes:(NSManagedObject *)managedObject
 {
-    [[NSNotificationCenter defaultCenter] addObserver:self
-                                             selector:@selector(handleDidSaveNotification:)
-                                                 name:NSManagedObjectContextDidSaveNotification
-                                               object:nil];
+	return (managedObject.entity.attributesByName[StoreVersion] != nil);
+}
 
+- (void)handleDidSaveNotification:(NSNotification *)notification
+{
+	NSSet *insertedObjects = notification.userInfo[NSInsertedObjectsKey];
+	for (NSManagedObject *insertedObject in insertedObjects)
+	{
+		if ([self entityHasRemoteStoreAttributes:insertedObject] == FALSE) {
+			// Only apply table operations on entities we expect to be managed by the mobile service
+			continue;
+		}
+		
+		NSString *tableName = insertedObject.entity.name;
+		NSDictionary *tableItem = [MSCoreDataStore tableItemFromManagedObject:insertedObject];
+		
+		MSSyncTable *syncTable = [self.client syncTableWithName:tableName];
+		[syncTable insert:tableItem completion:^(NSDictionary *item, NSError *error) {
+			if (error != nil) {
```

Comment:
Inserts, could in theory fall back to an updates, depending on the error (id already exists?)